internal/plugin/metrics_plugin.go

```diff
@@ -0,0 +1,224 @@
+		collectInterval time.Duration
+		shutdownFuncs   []shutdownFunc
+	}
+
+	// MetricsOption a functional option for `*Metrics`.
+	MetricsOption func(*Metrics) error
+)
+
+// NewMetrics is the constructor for the Metrics plugin.
+func NewMetrics(c config.Config, options ...MetricsOption) (*Metrics, error) {
+	m := Metrics{
+		producers:       make(map[model.MetricsSourceType]model.MetricsProducer),
+		exporters:       make(map[model.MetricsSourceType]model.Exporter),
+		conf:            c,
+		pipe:            nil,
+		collectInterval: fallbackCollectInterval,
+		shutdownFuncs:   make([]shutdownFunc, 0),
+	}
+
+	for _, opt := range options {
+		err := opt(&m)
+		if err != nil {
+			return nil, fmt.Errorf("failed to apply metrics plugin option: %w", err)
+		}
+	}
+
+	return &m, nil
+}
+
+// Init initializes and starts the plugin. Required for the `Plugin` interface.
+func (m *Metrics) Init(mp bus.MessagePipeInterface) error {
+	m.pipe = mp
+	ctx, cancel := context.WithCancel(mp.Context())
+	m.shutdownFuncs = append(m.shutdownFuncs, shutdownFunc(cancel))
+
+	m.discoverSources()
+
+	err := m.startExporters(ctx)
+	if err != nil {
+		return fmt.Errorf("could not start exporters: %w", err)
+	}
+
+	for srcType, producer := range m.producers {
+		slog.Info("Starting producer", "producer_type", srcType.String())
+		go m.runProducer(ctx, producer)
+	}
+
+	return nil
+}
+
+// Info about the plugin. Required for the `Plugin` interface.
+func (m *Metrics) Info() *bus.Info {
+	return &bus.Info{
+		Name: metricsInfo,
+	}
+}
+
+// Close about the plugin. Required for the `Plugin` interface.
+func (m *Metrics) Close() error {
+	for _, shutdown := range m.shutdownFuncs {
+		shutdown()
+	}
+
+	return nil
+}
+
+// Process an incoming Message Bus message in the plugin. Required for the `Plugin` interface.
+func (m *Metrics) Process(msg *bus.Message) {
+	switch msg.Topic {
+	case bus.MetricsTopic:
+		m.processMessage(msg)
+	case bus.OsProcessesTopic:
+		slog.Debug("OS Processes have been updated")
+	}
+}
+
+// Subscriptions returns the list of topics the plugin is subscribed to. Required for the `Plugin` interface.
+func (m *Metrics) Subscriptions() []string {
+	return []string{
+		bus.OsProcessesTopic,
+		bus.MetricsTopic,
+	}
+}
+
+// Dynamically populates the `[]*MetricsSource` of `*Metrics`. This is currently a one-to-one mapping of each data type
+// to its data sources, can later be changed to a one-to-many.
+func (m *Metrics) discoverSources() {
+	// Always initialized currently, should dynamically determine if we have a Prometheus sources or not.
+	// Data sources can also be passed in as options on plugin init.
+	if _, ok := m.producers[model.Prometheus]; !ok {
+		// Will always use the env var for the time being.
+		m.producers[model.Prometheus] = prometheus.NewScraper(make([]string, 0))
+	}
+}
+
+func (m *Metrics) startExporters(ctx context.Context) error {
+	for srcType := range m.producers {
+		// How do we know which exporter is used by which source?
```

Comment:
Oh yeah, you're right, we would want each Exporter type to be a singleton, I'll get that fixed. But I suppose we will have other exporters than OTel exporters, right?

Reply:
Discussed this at the office and the answer was 'yes', we want to support more exporters but each exporter of a given type should be a singleton, instead of instantiating a new exporter per source. This is now fixed.